test/helpers/kubectl.go

```diff
@@ -1408,6 +1432,450 @@ func (kub *Kubectl) Delete(filePath string) *CmdRes {
+func (kub *Kubectl) ValidateServicePlumbing(namespace, service string) error {
+	fullName := namespace + "/" + service
+
+	serviceObj, err := kub.GetService(namespace, service)
+	if err != nil {
+		return err
+	}
+
+	if serviceObj == nil {
+		return fmt.Errorf("%s service not found", fullName)
+	}
+
+	res := kub.Get(namespace, "endpoints "+service)
+	if !res.WasSuccessful() {
+		return fmt.Errorf("unable to retrieve endpoints %s: %s", fullName, res.OutputPrettyPrint())
+	}
+
+	if serviceObj.Spec.ClusterIP == v1.ClusterIPNone {
+		return nil
+	}
+
+	var endpointsObj v1.Endpoints
+	err = res.Unmarshal(&endpointsObj)
+	if err != nil {
+		return fmt.Errorf("unable to unmarshal endpoints %s: %s", fullName, err)
+	}
+
+	ciliumPods, err := kub.GetCiliumPods(GetCiliumNamespace(GetCurrentIntegration()))
+	if err != nil {
+		return err
+	}
+
+	var (
+		wg       sync.WaitGroup
+		podError = make(chan error, len(ciliumPods))
+	)
+	wg.Add(len(ciliumPods))
+	for _, ciliumPod := range ciliumPods {
+		go func(pod string) {
+			var err error
+			// The plumbing of Kubernetes services typically lags
+			// behind a little bit if Cilium was just restarted.
+			// Give this a thight timeout to avoid always failing.
+			timeoutErr := RepeatUntilTrue(func() bool {
+				err = kub.validateServicePlumbingInCiliumPod(fullName, pod, serviceObj, endpointsObj)
+				if err != nil {
+					ginkgoext.By("Checking service %s plumbing in cilium pod %s: %s", fullName, pod, err)
+				}
+				return err == nil
+			}, &TimeoutConfig{Timeout: 5 * time.Second, Ticker: 1 * time.Second})
+			if err != nil {
+				podError <- err
+			} else if timeoutErr != nil {
+				podError <- timeoutErr
+			}
+			wg.Done()
+		}(ciliumPod)
+	}
+	wg.Wait()
+
+	select {
+	case err := <-podError:
+		return err
+	default:
+	}
+
+	return nil
+}
+
+// ValidateKubernetesDNS validates that the Kubernetes DNS server has been
+// deployed correctly and can resolve DNS names. The following validations are
+// done:
+//  - The Kuberentes DNS deployment has at least one replica
+//  - All replicas are up-to-date and ready
+//  - All pods matching the deployment are represented by a CiliumEndpoint with an identity
+//  - The kube-system/kube-dns service is correctly pumbed in all Cilium agents
+//  - The service "default/kubernetes" can be resolved via the KubernetesDNS
+//    and the IP returned matches the ClusterIP in the service
+func (kub *Kubectl) ValidateKubernetesDNS() error {
+	// The deployment is always validated first and not in parallel. There
+	// is no point in validating correct plumbing if the DNS is not even up
+	// and running.
+	ginkgoext.By("Checking if deployment is ready")
+	_, err := kub.DeploymentIsReady(KubeSystemNamespace, "kube-dns")
+	if err != nil {
+		_, err = kub.DeploymentIsReady(KubeSystemNamespace, "coredns")
+		if err != nil {
+			return err
+		}
+	}
+
+	var (
+		wg       sync.WaitGroup
+		errQueue = make(chan error, 3)
+	)
+	wg.Add(3)
```

Comment:
We recently started using https://godoc.org/golang.org/x/sync/errgroup , since we are returning first error we get it seems to be a good place to use.